### DISCUSSION TITLE
Segfault when creating mesh entity

### DISCUSSION THREAD

**Discussion by ComradeMashkov** [pelicanmapping/rocky#38] [Q&A / Support]
I have tried to load line entity and it works perfectly, but when loading mesh entity I have segmentation fault (rdemo also fails). I guess it is related to shaders problem, but I don't know exactly why mesh loading gives me segfault. Currently I can't reach debugger in Visual Studio (my personal problem with laptop), so if you can reproduce this problem I will appreciate it! Here is the code:

```cpp
#include <rocky/vsg/Application.h>
#include <rocky/vsg/Mesh.h>

#ifdef ROCKY_HAS_MBTILES
#include <rocky/MBTilesImageLayer.h>
#endif

#include <iostream>

int main(int argc, char** argv)
{
    rocky::Application app;

    #ifdef ROCKY_HAS_MBTILES
    auto layer = rocky::MBTilesImageLayer::create();
    layer->setURI("OAM-World-1-8-min-J80.mbtiles");
    app.mapNode->map->layers().add(layer);

    if (layer->status().failed())
    {
        std::cerr << layer->status().toString() << std::endl;
    }
    #endif

    // World
    auto entity = app.entities.create();

    auto& xform = app.entities.emplace<rocky::Transform>(entity);
    xform.setPosition(rocky::GeoPoint(rocky::SRS::WGS84, 0, 0, 50000));

    // Mesh
    auto mesh_entity = app.entities.create();
    auto& mesh = app.entities.emplace<rocky::Mesh>(mesh_entity);

    auto mesh_xform = rocky::SRS::WGS84.to(rocky::SRS::WGS84.geocentricSRS());
    const double step = 2.5;
    const double alt_mesh = 0.0;
    for (double lon = 0.0; lon < 35.0; lon += step)
    {
        for (double lat = 15.0; lat < 35.0; lat += step)
        {
            vsg::dvec3 v1, v2, v3, v4;
            mesh_xform(vsg::dvec3{ lon, lat, alt_mesh }, v1);
            mesh_xform(vsg::dvec3{ lon + step, lat, alt_mesh }, v2);
            mesh_xform(vsg::dvec3{ lon + step, lat + step, alt_mesh }, v3);
            mesh_xform(vsg::dvec3{ lon, lat + step, alt_mesh }, v4);

            mesh.add({ {v1, v2, v3} });
            mesh.add({ {v1, v3, v4} });
        }
    }

    mesh.style = { {1, 0.4, 0.1, 0.75}, 32.0f, 1e-7f };
    mesh.writeDepth = true;

    return app.run();
}
```

**Rocky**: [master](https://github.com/pelicanmapping/rocky/commit/75f6555f2bcd0ae7e11dd45088c9d011a0d280bc)
**Platform**: MSVC 17 2022 Win10
I have no NV extension for shaders enabled on laptop and it crashes. On my PC I have this extension enabled and it works (though I ran rdemo a couple of weeks ago).

**Comment by ComradeMashkov** [comment]
Ok, I have reached debugger in Visual Studio and here is what I've found out:

<details>

<summary> Error </summary>

![image](https://github.com/pelicanmapping/rocky/assets/117441107/d007a92d-ae28-4b42-83f5-6678dfb88391)

</details>

<details>

<summary> Call Stack </summary>

![image](https://github.com/pelicanmapping/rocky/assets/117441107/c8f5472d-8a32-4ab0-ade0-a687b535515d)

</details>

**Comment by gwaldron** [comment]
I'm not able to reproduce this issue. I agree that some Vulkan objects failed to compile, and it could be a missing shader.
Are there any errors or warnings on the command line?

**Reply by ComradeMashkov** [reply]
No, just "Segmentation fault". The debugger gives me the output I've shown above.

**Reply by ComradeMashkov** [reply]
As far as I can see old GPUs (such as GTX 1050, 1060, 1070 and 1080) do not support barycentric coords either. Does this code work for you if you will disable NV shader barycentric extension?

**Reply by ComradeMashkov** [reply]
The problem is definitely in rocky.mesh.frag and rocky.mesh.vert files. If I load a program without any of these files it gives me the desired mesh but without any texture or shader.

**Comment by ComradeMashkov** [comment]
I managed to temporarary fix the problem by editing rocky.mesh.frag code. It works but it renders only the cube, not the vertices (lines on the cube from the code I've shown above).
```frag
#version 450
#pragma import_defines(USE_MESH_TEXTURE)

layout(location = 1) in vec2 uv;

// inter-stage interface block
struct Varyings {
    vec4 color;
    float wireframe;
    vec3 barycentric;
};
layout(location = 2) flat in Varyings vary;

// outputs
layout(location = 0) out vec4 out_color;

// unis
#ifdef USE_MESH_TEXTURE
layout(binding=6) uniform sampler2D mesh_texture;
#endif

vec3 calculateBarycentric(vec2 p, vec2 a, vec2 b, vec2 c)
{
    vec2 v0 = b - a, v1 = c - a, v2 = p - a;
    float d00 = dot(v0, v0);
    float d01 = dot(v0, v1);
    float d11 = dot(v1, v1);
    float d20 = dot(v2, v0);
    float d21 = dot(v2, v1);
    float denom = d00 * d11 - d01 * d01;
    float v = (d11 * d20 - d01 * d21) / denom;
    float w = (d00 * d21 - d01 * d20) / denom;
    float u = 1.0 - v - w;
    return vec3(u, v, w);
}


void main()
{
    out_color = vary.color;

#ifdef USE_MESH_TEXTURE
    out_color.rgb *= texture(mesh_texture, uv).rgb;
#endif

    if (vary.wireframe > 0.0)
    {
        vec2 vertex0 = vec2(0.0, 0.0);
        vec2 vertex1 = vec2(1.0, 0.0);
        vec2 vertex2 = vec2(0.5, 1.0);

        float b = min(vary.barycentric.x, min(vary.barycentric.y, vary.barycentric.z)) * vary.wireframe;
        out_color.rgb = mix(vec3(1, 1, 1), out_color.rgb, clamp(b, 0.85, 1.0));
    }
}
```

<details>

<summary> Picture </summary>

![image](https://github.com/pelicanmapping/rocky/assets/117441107/16d919cb-92bd-46ab-8988-3eee9dc74398)

</details>

**Comment by gwaldron** [comment]
Do you have the latest code? Wasn't this fixed in 8e96f112e ?

**Reply by ComradeMashkov** [reply]
I have master version of the Rocky. Without https://github.com/pelicanmapping/rocky/commit/8e96f112e903054e05173599494bf297dcec5325 I couldn't even run executable file. However, mesh shaders are failed to compile because I don't have NV barycentric extension. So I decided to add barycentric coordinates calculation in .frag file.
